### PR TITLE
ZQS 1104 New pauli operator data structure

### PR DESCRIPTION
**Pull request by laurgao**
## Description

Migrate our code from using openfermion's QubitOperator and IsingOperator to our own PauliTerm and PauliSum.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

File: src/orquestra/quantum/measurements/measurements.py [hunk 1]
```diff
@@ -285,8 +286,9 @@ def get_distribution(self) -> MeasurementOutcomeDistribution:
 
         return MeasurementOutcomeDistribution(distribution)
 
+    # Union[PauliSum, PauliTerm]
```
Comment (mstechly) [inline src/orquestra/quantum/measurements/measurements.py:289]: I don't think this is necessary here?

File: src/orquestra/quantum/openfermion/linalg/sparse_tools.py [hunk 2]
```diff
@@ -144,7 +147,73 @@ def jordan_wigner_sparse(fermion_operator, n_qubits=None):
     return sparse_operator
 
 
-def qubit_operator_sparse(qubit_operator, n_qubits=None):
+def pauli_sum_sparse(qubit_operator: PauliSum, n_qubits=None):
```
Comment (mstechly) [inline src/orquestra/quantum/openfermion/linalg/sparse_tools.py:150]: If we need this (I'm not sure if we do 🤔 ) it should live in the `operators` dir, not in `openfermion`.

File: src/orquestra/quantum/openfermion/utils/operator_utils.py [hunk 3]
```diff
@@ -138,6 +139,12 @@ def count_qubits(operator):
     Raises:
        TypeError: Operator of invalid type.
     """
+    # Handle PauliSum
+    if isinstance(operator, PauliSum):
```
Comment (mstechly) [inline src/orquestra/quantum/openfermion/utils/operator_utils.py:143]: I think we won't need this anymore in "post openfermion" world.

File: src/orquestra/quantum/wip/operators/__init__.py [hunk 4]
```diff
@@ -1 +1 @@
-from ._pauli_operators import PauliSum, PauliTerm
+from ._pauli_operators import PauliRepresentation, PauliSum, PauliTerm
```
Comment (mstechly) [inline src/orquestra/quantum/wip/operators/__init__.py:1]: Honestly I would rename `PauliRepresentation` to `PauliOperator`.
I don't like `PauliRepresentation` – thoughts @dexter2206 ?

File: tests/orquestra/quantum/estimation/estimation_test.py [hunk 5]
```diff
@@ -17,18 +17,18 @@
...
 
 
 class TestEstimatorUtils:
     @pytest.fixture()
     def frame_operators(self):
         operators = [
-            IsingOperator("2 [Z1 Z2]"),
-            IsingOperator("[Z0 Z3]"),
-            IsingOperator("-1 [Z2]"),
+            PauliTerm.from_str("2*Z1*Z2"),
```
Comment (mstechly) [inline tests/orquestra/quantum/estimation/estimation_test.py:29]: Hmm...
Should we keep an option to create those PauliTerms/Sums from openfermion strings? Just for compatibility reasons 🤔 ?
CC: @dexter2206

File: tests/orquestra/quantum/measurements/measurements_test.py [hunk 6]
```diff
@@ -348,7 +352,11 @@ def test_get_expectation_values_from_measurements_with_bessel_correction(self):
         measurements = Measurements(
             [(0, 1, 0), (0, 1, 0), (0, 0, 0), (1, 0, 0), (1, 1, 1)]
         )
-        ising_operator = IsingOperator("10[] + [Z0 Z1] - 15[Z1 Z2]")
+        ising_operator = (
+            PauliTerm("I0", 10)
+            + PauliTerm.from_str("1*Z0*Z1")
```
Comment (mstechly) [inline tests/orquestra/quantum/measurements/measurements_test.py:357]: I'm wondering if we shouldn't allow for:
`PauliTerm("1*Z0*Z1")` as well?
I mean without `from_str`?
Maybe not a good idea, not sure about it 🤔

File: src/orquestra/quantum/openfermion/linalg/sparse_tools.py (hunk 2 above)
Comment (laurgao) [reply src/orquestra/quantum/openfermion/linalg/sparse_tools.py:150]: This is used for getting the expectation value in `backend`. We will move it to the right place in a later PR when we move the operators out of `wip` and get rid of openfermion.

File: src/orquestra/quantum/openfermion/utils/operator_utils.py (hunk 3 above)
Comment (laurgao) [reply src/orquestra/quantum/openfermion/utils/operator_utils.py:143]: Same as above

File: tests/orquestra/quantum/estimation/estimation_test.py (hunk 5 above)
Comment (laurgao) [reply tests/orquestra/quantum/estimation/estimation_test.py:29]: Also will do in separate PR

**Review by dexter2206 — APPROVED** [review]

**Pull request merged**